### ISSUE TITLE
`azd env set` does not check for case sensitivity

### ISSUE THREAD

**Issue by markweitzel** [Azure/azure-dev#289]

**Comment by savannahostrowski** [comment]
Is this still reproduceable?

**Comment by rajeshkamal5050** [comment]
@v-xuto can you try to repro?

**Comment by v-xuto** [comment]
@rajeshkamal5050 This issue can be reproduced, and we will try to fix it.

**Comment by vhvb1989** [comment]
@v-xuto , before you start a PR for this, can you share what is the fix you would be doing?  Since there are 3 options on the issue, we need to make sure about including @ellismg and @kristenwomack for thumbs up :)

**Comment by NanaXiong00** [comment]
@vhvb1989 We will choose the "force upper case" method to fix this issue.

**Comment by vhvb1989** [comment]
@NanaXiong00 , in general .env files are case sensitive.  Developers might be familiar with this concept as it is well known in all languages when using .env files.  Maybe forcing upper case could make azd to work differently than the universal expectation.

**Comment by rajeshkamal5050** [comment]
@vhvb1989 is there anything which needs to be addressed here? should we close this as wont-fix?

**Comment by vhvb1989** [comment]
@rajeshkamal5050 , I thing the approach of `warn if the same value is present but in mixed case` would be enough here

**Comment by kristenwomack** [comment]
> [@rajeshkamal5050](https://github.com/rajeshkamal5050) , I thing the approach of `warn if the same value is present but in mixed case` would be enough here

I like this solution the most 👆

**Comment by v-xuto** [comment]
> [@rajeshkamal5050](https://github.com/rajeshkamal5050) , I thing the approach of `warn if the same value is present but in mixed case` would be enough here

Okay, thank you for your suggestions. We will use this approach to fix.

**Comment by v-xuto** [comment]
@rajeshkamal5050, @vhvb1989 PR https://github.com/Azure/azure-dev/pull/4547 is ready, please help review.

**Comment by weikanglim** [comment]
Just double checking on our UX requirements here with @vhvb1989 and @kristenwomack :

Would we want to print a warning message, and then ask the user to confirm?

In the current PR being proposed, we would print the warning message then continue -- I think this does meet the requirement of informing the user that this is happening, but it may be too late as the value is already saved. We don't have a straightforward `undo` command today, so I wonder if an upfront confirmation is better.

**Comment by vhvb1989** [comment]
> Just double checking on our UX requirements here with [@vhvb1989](https://github.com/vhvb1989) and [@kristenwomack](https://github.com/kristenwomack) :
> 
> Would we want to print a warning message, and then ask the user to confirm?
> 
> In the current PR being proposed, we would print the warning message then continue -- I think this does meet the requirement of informing the user that this is happening, but it may be too late as the value is already saved. We don't have a straightforward `undo` command today, so I wonder if an upfront confirmation is better.

I don't think we need a confirmation.  The warning should be about there is another env var with the same name but with a different casing.
For example, I can run `azd env set foo bar` with no warning. Then if I run `azd env set Foo bar`, there should be a warning like `Warning: There's `foo` in the environment as well`.
Then, if I run `azd env set FOO bar`, I would love to see `Warning: There is `foo` and `Foo` in the environment`.

This warning is just to help user to know about typos, but there no override or data lost.  So, no confirmation is required.

**Comment by weikanglim** [comment]
Just wondering, in the scenario that a user provided `AZURE_Location` instead of `AZURE_LOCATION`, and `AZURE_LOCATION` exists:

- Is more likely that the user intended to set `AZURE_Location` and `AZURE_LOCATION` separately?
- Or is it more likely that the user mistyped `AZURE_LOCATION` as `AZURE_Location`?

My guesstimate is that 9 out of 10 times, they meant to set `AZURE_LOCATION`. A UX design that prompts the user for confirmation would be helpful for that. We could even explicitly ask them what they would like to do -- set as `AZURE_Location`, `AZURE_LOCATION`, or cancel. This way, we balance the design to satisfy both needs.

I personally feel that this type of UX interaction handles the user intention the best, although it does come at a cost of an extra prompt. I'd defer the overall UX decision to @SophCarp and the UX team

**Comment by vhvb1989** [comment]
Adding a prompt is also a breaking change for anyone calling `azd env ser x y` from CI/CD (or any script).  Even if we add a default choice to keep the old behavior as the default, folks don't need to use `--no-promt` for `azd env set x y`, which would stop the flow.

**Comment by weikanglim** [comment]
That's a great observation!

I did think about that as well, here's where I currently stand on programmatic access for commands:

`--no-prompt` is designed specifically on all commands to handle the programmatic access scenarios, and setting `--no-prompt` would be the design we should continue promoting. I'm not sure if this aligns with where @ellismg was at when we started with this pattern. In an ideal world, I could think about making this flag a little shorter, like a `-y` or `--yes` -- to say "yes" to defaults.

I personally don't think we should consider adding a prompt to an existing command as a breaking change. For this particular scenario, I can't imagine many cases where users would want to programmatically set `AZURE_location` and `AZURE_LOCATION` at the same time, so I would avoid generalizing the concern at this level.

From where I stand, I don't feel happy about the current proposal -- the UX is jarring personally. As a design guideline, I would discourage the use of warning messages except in cases where the source of "error" is an invariant outside the system, that can't be fixed.

I'd still propose the following:

## Option 1: Message for case sensitivity

I'd recommend structuring the message as a Hint rather than a Warning.

- Currently we have:

![Image](https://github.com/user-attachments/assets/7568ccde-1279-457e-a8ab-4b505db01b2c)

- My proposal would be:

Hint: Did you mean to set 'foo' instead of 'Foo'?

## Option 2: Confirmation for case sensitivity

My proposal would be:

? Add 'foo' as an environment value, while 'Foo' exists? (Y/n)

I understand that these are details that we may not think matter, but it's a great exercise for us to walk through and come up with clear design language that aids us from making better informed decisions in the future. To that end, I'm happy to continue discussion further until we reach some sort of consensus.

**Comment by v-xuto** [comment]
@vhvb1989 Any ideas on the [above comments](https://github.com/Azure/azure-dev/issues/289#issuecomment-2477152902) by Wei's?

**Comment by vhvb1989** [comment]
@weikanglim , I like the `hint` approach.  But I don't hate the `Warning` XD

Maybe b/c of my old times with C++, but, I think about `warnings` as the union of hints, notes, remarks, or whatever is not considered an error for the compiler, but it is recognized as a potential `unwanted scenario`.

I think about scenarios where customers can create a script that contains some azd commands.  When running the full script, it is common to filter the output for either `Errors` or `Warnings` to spot issues.  Maybe adding something like a `Hint` would not be expected right away from someone looking at a long script output.

For non-script cases, like when customers run azd directly, I don't care too much about using `hint` v/s `warning`

**Comment by Menghua1** [comment]
This issue has been fixed by https://github.com/Azure/azure-dev/pull/4547, so I will close this issue. If you still have questions about it, you can reopen this issue.
